text/0192-remove-constructs-compat.md

```diff
@@ -0,0 +1,1095 @@
+all constructs must use the same version of the `Construct` base class.
+
+As long as all libraries in a closure take a __peer dependency__ on a compatible
+version of `constructs`, the npm package manger will include a single copy of
+the library, and therefore all constructs will derive from the same `Construct`
+(and more importantly, accept the same `Construct` for `scope`).
+
+Practically this means that we can never introduce a major version of
+`constructs` because any major version will require a new major version of all
+CDKs, and that is impossible to require or coordinate given the decentralized
+nature of the ecosystem.
+
+We propose to take a commitment to __never introduce breaking changes in
+"constructs"__. This implies that we will never introduce another major version.
+To symbolize that to users, we will use the major version `10.x`.
+
+## 01-BASE-TYPES
+
+Once `construct-compat.ts` is removed from `@aws-cdk/core`, all CDK code (both
+the framework itself and user code) would need to be changed to use the types
+from `constructs`.
+
+Since the APIs are similar in almost all cases, this is a simple mechanical
+change as shown in the [release notes](#release-notes).
+
+### What can we do on 1.x?
+
+The main concern for the CDK codebase is maintaining this change alongside a 1.x
+branch until we switch over to 2.x. Since this change includes modifications to
+all `import` sections in almost all of our files, merge conflicts are imminent.
+
+To reduce these costs, we propose to modify the `scope` argument on all  1.x to
+accept `constructs.Construct` instead of `core.Construct`.
+
+This will provide the following benefits (enforced by an `awslint` rule):
+
+1. The `import` statement for `import { Construct } from 'constructs'` would
+   already exist which will reduce merge conflicts.
+2. It will unlock composition of framework constructs into other domains (e.g.
+   it will be possible to pass an AWS CDK L2 to a terraform stack).
+
+Note that we will *not* change the base classes to `constructs.Construct`
+because it is technically (and practically) a breaking API change (we must maintain
+the invariant that "`s3.Bucket` is a `core.Construct`".
+
+The remaining change in 2.x will be to update any base classes to use
+`constructs.Construct` but this is actually not very prevalent in the framework
+code because the majority of constructs actually extend `core.Resource`.
+
+Alternatives considered:
+
+- **Do nothing in 1.x**: will incur an ongoing maintenance cost of the 1.x -> 2.x
+  merge conflicts.
+- **Automatic merge resolution and import organization**: requires research and
+  development, not necessarily worth it.
+
+## 10-CONSTRUCT-NODE
+
+As we realize that we won't be able to release additional major versions of the
+"constructs" in order to ensure interoperability between domains is always
+possible, we need to closely examine the API of this library.
```

Comment:
```suggestion
Since we won't be able to release additional major versions of the
"constructs" library (in order to ensure interoperability between domains is always
possible), we need to closely examine the API of this library.
```